internal/logger/telemetry_store_watermill.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024 Stacklok, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logger
+
+import (
+	"fmt"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/rs/zerolog"
+
+	"github.com/stacklok/minder/internal/engine"
+	minderv1 "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+)
+
+// TelemetryStoreWMMiddleware is a Watermill middleware that
+// logs the relevant telemetry data.
+type TelemetryStoreWMMiddleware struct {
+	l *zerolog.Logger
+}
+
+// NewTelemetryStoreWMMiddleware returns a new TelemetryStoreWMMiddleware.
+func NewTelemetryStoreWMMiddleware(l *zerolog.Logger) *TelemetryStoreWMMiddleware {
+	return &TelemetryStoreWMMiddleware{l: l}
+}
+
+// TelemetryStoreMiddleware is a Watermill middleware that
+// logs the relevant telemetry data.
+func (m *TelemetryStoreWMMiddleware) TelemetryStoreMiddleware(h message.HandlerFunc) message.HandlerFunc {
+	return func(msg *message.Message) ([]*message.Message, error) {
+		if err := m.Log(msg); err != nil {
+			return nil, fmt.Errorf("error logging telemetry: %w", err)
+		}
+
+		return h(msg)
+	}
+}
+
+// Log logs the relevant telemetry data.
+func (m *TelemetryStoreWMMiddleware) Log(msg *message.Message) error {
+	inf, err := engine.ParseEntityEvent(msg)
+	if err != nil {
+		return fmt.Errorf("error unmarshalling payload: %w", err)
+	}
+
+	typ := inf.Type
+
+	ts := &TelemetryStore{
+		Project: inf.ProjectID.String(),
+	}
+
+	repoID, artID, prID := inf.GetEntityDBIDs()
+
+	// In minder this would be the entity
+	switch typ {
+	case minderv1.Entity_ENTITY_UNSPECIFIED:
+		return fmt.Errorf("unspecified entity type")
+	case minderv1.Entity_ENTITY_BUILD_ENVIRONMENTS:
+		return fmt.Errorf("build environments not supported")
+	case minderv1.Entity_ENTITY_REPOSITORIES:
+		ts.Resource = repoID.String()
+	case minderv1.Entity_ENTITY_ARTIFACTS:
+		ts.Resource = artID.UUID.String()
+	case minderv1.Entity_ENTITY_PULL_REQUESTS:
+		ts.Resource = prID.UUID.String()
+	}
+
+	ctx := ts.WithTelemetry(msg.Context())
+	msg.SetContext(ctx)
+
+	ts.Record(m.l.Info()).Str("event", msg.UUID).
```

Comment:
I think we might not want to call `Send()` on the Zerolog.Event until _after_ `h` is run, because `h` might want to augment the event with additional details (for example, the names of rules that were run).